kernel/src/expressions/scalars.rs

```diff
@@ -999,4 +1068,187 @@ mod tests {
+        assert!(matches!(scalar, Scalar::Map(_)));
+
+        // Verify the data type
+        let expected_map_type = MapType::new(DataType::STRING, DataType::INTEGER, false);
+        assert_eq!(scalar.data_type(), DataType::from(expected_map_type));
+
+        // Extract the MapData and verify contents
+        if let Scalar::Map(map_data) = scalar {
+            let pairs = map_data.pairs();
+            assert_eq!(pairs.len(), 2);
+
+            // Check that both expected pairs are present
+            let has_key1 = pairs.iter().any(|(k, v)| {
+                matches!(k, Scalar::String(s) if s == "key1") && matches!(v, Scalar::Integer(42))
+            });
+            let has_key2 = pairs.iter().any(|(k, v)| {
+                matches!(k, Scalar::String(s) if s == "key2") && matches!(v, Scalar::Integer(100))
+            });
+
+            assert!(has_key1, "Missing key1 -> 42 pair");
+            assert!(has_key2, "Missing key2 -> 100 pair");
+        } else {
+            panic!("Expected Map scalar");
+        }
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_hashmap_conversion_with_nullable_values() -> DeltaResult<()> {
+        // Create a HashMap with string keys and optional integer values
+        let mut map = HashMap::new();
+        map.insert("key1".to_string(), Some(42i32));
+        map.insert("key2".to_string(), None);
+        map.insert("key3".to_string(), Some(100i32));
+
+        // Convert HashMap to Scalar
+        let scalar = Scalar::try_from(map)?;
+
+        // Verify the scalar is of Map type
+        assert!(matches!(scalar, Scalar::Map(_)));
+
+        // Verify the data type (should have value_contains_null = true)
+        let expected_map_type = MapType::new(DataType::STRING, DataType::INTEGER, true);
+        assert_eq!(scalar.data_type(), DataType::from(expected_map_type));
+
+        // Extract the MapData and verify contents
+        if let Scalar::Map(map_data) = scalar {
+            let pairs = map_data.pairs();
+            assert_eq!(pairs.len(), 3);
+
+            // Check that all expected pairs are present
+            let has_key1 = pairs.iter().any(|(k, v)| {
+                matches!(k, Scalar::String(s) if s == "key1") && matches!(v, Scalar::Integer(42))
+            });
+            let has_key2 = pairs.iter().any(|(k, v)| {
+                matches!(k, Scalar::String(s) if s == "key2") && matches!(v, Scalar::Null(_))
+            });
+            let has_key3 = pairs.iter().any(|(k, v)| {
+                matches!(k, Scalar::String(s) if s == "key3") && matches!(v, Scalar::Integer(100))
+            });
+
+            assert!(has_key1, "Missing key1 -> 42 pair");
+            assert!(has_key2, "Missing key2 -> null pair");
+            assert!(has_key3, "Missing key3 -> 100 pair");
+
+            // Verify the map type allows null values
+            assert!(map_data.map_type().value_contains_null());
+        } else {
+            panic!("Expected Map scalar");
+        }
```

Comment:
similar here, return early with a `let Scalar::Map(map_data) = scalar else {panic}`
```suggestion
       let Scalar::Map(map_data) = scalar else {
             panic!("Expected Map scalar");
       }
      let pairs = map_data.pairs();
      assert_eq!(pairs.len(), 3);

      // Check that all expected pairs are present
      let has_key1 = pairs.iter().any(|(k, v)| {
          matches!(k, Scalar::String(s) if s == "key1") && matches!(v, Scalar::Integer(42))
      });
      let has_key2 = pairs.iter().any(|(k, v)| {
          matches!(k, Scalar::String(s) if s == "key2") && matches!(v, Scalar::Null(_))
      });
      let has_key3 = pairs.iter().any(|(k, v)| {
          matches!(k, Scalar::String(s) if s == "key3") && matches!(v, Scalar::Integer(100))
      });

      assert!(has_key1, "Missing key1 -> 42 pair");
      assert!(has_key2, "Missing key2 -> null pair");
      assert!(has_key3, "Missing key3 -> 100 pair");

      // Verify the map type allows null values
      assert!(map_data.map_type().value_contains_null());
```